### PR TITLE
feat: implement exponential retry mechanism for handling network errors

### PR DESCRIPTION
**Pull request by SKannaniOS**
## Description

- This PR removes the existing retry mechanism for network errors and replaces it with an **exponential retry mechanism with jitter**. Now, retries will occur exponentially for every **4xx** and **5xx** status code (except 400 and 404). If the network is unavailable, retries will continue with a one-second delay.
- The maximum limit is 5 minutes; once this limit is reached, the retry process will start over from the beginning.
- Example sequence:
```
6s, 8s, 15s, 28s, 1m 35s, 2m 48s, 3m 49s, 5m 0s,
4s, 6s, 24s, 31s..
```

**NOTE**: No event is deleted in this process.

### CODE REVIEW

File: Sources/Classes/RSCloudModeManager.m [hunk 1]
```diff
@@ -60,20 +63,24 @@ - (void) startCloudModeProcessor {
             [strongSelf->lock unlock];
             [RSLogger logDebug:[[NSString alloc] initWithFormat:@"RSCloudModeManager: CloudModeProcessor: cloudModeSleepCount: %d", sleepCount]];
             sleepCount += 1;
+
```
Comment (1abhishekpandey) [inline Sources/Classes/RSCloudModeManager.m:66]: Please add the logic to `reset` the backoff when events are sent successfully:
```
[self-> backOff reset];
```

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: Sources/Classes/RSExponentialBackOff.m [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * Function will calculate the next delay value in seconds
+ */
+- (int)nextDelay {
+    int delay = (int)pow(2, _attempt);
+    _attempt = _attempt + 1;
+    
+    int jitter = arc4random_uniform((delay + 1));
+    
+    int exponentialDelay = _initialDelay + delay + jitter;
+    exponentialDelay = MIN(exponentialDelay, _maximumDelay);
```
[the remark this answers is not in the document]
Comment (SKannaniOS) [reply Sources/Classes/RSExponentialBackOff.m:44]: arc4random_uniform - always return the random value from 0 to < upper bound value. To include the calculated delay value, 1 added.

File: Sources/Classes/RSCloudModeManager.m [hunk 3]
```diff
@@ -60,20 +63,24 @@ - (void) startCloudModeProcessor {
             [strongSelf->lock unlock];
             [RSLogger logDebug:[[NSString alloc] initWithFormat:@"RSCloudModeManager: CloudModeProcessor: cloudModeSleepCount: %d", sleepCount]];
             sleepCount += 1;
+            
             if(response == nil) {
```
[the remark this answers is not in the document]
Comment (SKannaniOS) [reply Sources/Classes/RSCloudModeManager.m:67]: As per the RSCloudModeManager code logic, other than "WRONG_WRITE_KEY", every network error code is retryable and no messages are dropped from local DB. So please guide on the logic & what needs to be implemented.

[the remark this answers is not in the document]
Comment (saikumarrs) [reply Sources/Classes/RSCloudModeManager.m:67]: Here is a simplistic overview:
2XX - success and proceed
5xx and 429 - retry
others - abort and proceed

However, looks like we want to address this in v2 SDKs. So, resolve this thread unless you have any further comments.

[the remark this answers is not in the document]
Comment (SKannaniOS) [reply Sources/Classes/RSCloudModeManager.m:69]: As per the last discussion, the PR has the implementation of exponential retry mechanism.

**Review by saikumarrs — APPROVED** [review]

**Review by ChryssaAliferi — APPROVED** [review]

**Review by 1abhishekpandey — APPROVED** [review]

**Pull request merged**